# Goodreads-books


<img src="Images/book1.jpg">

The basic idea behind analysing the Goodreads dataset is to get a fair idea about the relationships between the multiple attributes a book might have, such as:the aggregrate rating of each book, the trend of the authors over the years and books with numerous languages. With over a hundred thousand ratings, there are books which just tend to become popular as each day seems to pass.

We've always conisdered the magical persona books seem to hold, and with this notebook, we step out on a journey to see what kind of books really drives people to read in this era of modern smart devices.

With such a vast, overwhelming number of factors, we'll go over such demographics:

* Does any relationship lie between ratings and the total ratings given?
* Where do majority of the books lie, in terms of ratings - Does reading a book really bring forth bias for the ratings?
* Do authors tend to perform same over time, with all their newer books? Or do they just fizzle out.
* Do number of pages make an impact on reading styles, ratings and popularity?
* Can books be recommended based on ratings? Is that a factor which can work?

## 1).Import Module

* **Packages Imported** :-

    1. sys: access to system parameters
    2. numpy : for numerical computation
    3. pandas: for data manipulation and analysis
    4. matplotlib: plotting library
    5. seaborn: plotting library based on matplotlib
    6. sklearn: machine learning library

In [1]:
# load packages
import numpy as np
import pandas as pd

In [2]:
#import visual modules
import seaborn as sns
import matplotlib.pyplot as plt
sns.set_style("whitegrid")

In [3]:
## version of packages used
print(f'Numpy version: {np.__version__}')
print(f'Pandas version: {pd.__version__}')
#print(f'Matplotlib version: {plt.__version__}')
print(f'Seaborn version: {sns.__version__}')

Numpy version: 1.15.4
Pandas version: 0.23.4
Seaborn version: 0.9.0


## 2).The Data

**The data contain following information :-**

* **bookIDA :** unique Identification number for each book.
* **title :** The name under which the book was published.
* **authors :** Names of the authors of the book. Multiple authors are delimited with/.
* **average_rating :** The average rating of the book received in total.
* **isbn :** Another unique number to identify the book, the International Standard Book Number.
* **isbn13 :** A 13-digit ISBN to identify the book, instead of the standard 11-digit ISBN.
* **language_code :** Helps understand what is the primary language of the book. For instance, eng is standard for English.
* **num_pages :** Number of pages the book contains.
* **ratings_count :** Total number of ratings the book received.
* **text_reviews_count :** Total number of written text reviews the book received.
* **publication_date :** Date when the book was first published.
* **publisher :** The name of the publisher. 


**Read the data from the books.csv file in Books folder**

In [4]:
#save the data from books.csv in books dataframe
books=pd.read_csv("Books/books.csv",error_bad_lines=False)

b'Skipping line 3350: expected 12 fields, saw 13\nSkipping line 4704: expected 12 fields, saw 13\nSkipping line 5879: expected 12 fields, saw 13\nSkipping line 8981: expected 12 fields, saw 13\n'


Using head function to display the data in books

In [5]:
#displaying first 5 rows from the books datasset
books.head()

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   
2       4  Harry Potter and the Chamber of Secrets (Harry...   
3       5  Harry Potter and the Prisoner of Azkaban (Harr...   
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors  average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré            4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré            4.49  0439358078  9780439358071   
2                J.K. Rowling            4.42  0439554896  9780439554893   
3  J.K. Rowling/Mary GrandPré            4.56  043965548X  9780439655484   
4  J.K. Rowling/Mary GrandPré            4.78  0439682584  9780439682589   

  language_code    num_pages  ratings_count  text_reviews_count  \
0           eng          652        2095690               27591   
1           eng          870        2153167               29221   
2           eng          352           6333                 244   
3           eng          435        2339585               36325   
4           eng         2690          41428                 164   

  publication_date        publisher  
0        9/16/2006  Scholastic Inc.  
1         9/1/2004  Scholastic Inc.  
2        11/1/2003       Scholastic  
3         5/1/2004  Scholastic Inc.  
4        9/13/2004       Scholastic

In [6]:
#using shape funtion to get the number of rows and column of the data.
books.shape

(11123, 12)

There are 11123 rows and 12 columns in books dataframe


In [7]:
#Using describe function to view the statistical details like percentile ,mean,std etc.
books.describe()

bookID  average_rating        isbn13     num_pages  \
count  11123.000000    11123.000000  1.112300e+04  11123.000000   
mean   21310.856963        3.934075  9.759880e+12    336.405556   
std    13094.727252        0.350485  4.429758e+11    241.152626   
min        1.000000        0.000000  8.987060e+09      0.000000   
25%    10277.500000        3.770000  9.780345e+12    192.000000   
50%    20287.000000        3.960000  9.780582e+12    299.000000   
75%    32104.500000        4.140000  9.780872e+12    416.000000   
max    45641.000000        5.000000  9.790008e+12   6576.000000   

       ratings_count  text_reviews_count  
count   1.112300e+04        11123.000000  
mean    1.794285e+04          542.048099  
std     1.124992e+05         2576.619589  
min     0.000000e+00            0.000000  
25%     1.040000e+02            9.000000  
50%     7.450000e+02           47.000000  
75%     5.000500e+03          238.000000  
max     4.597666e+06        94265.000000

~From the data we got various statistical information like max averge_rating is 5. which means the rating is out of 5.

In [8]:
#using the info function to get the concise summary of the dataframe
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11123 entries, 0 to 11122
Data columns (total 12 columns):
bookID                11123 non-null int64
title                 11123 non-null object
authors               11123 non-null object
average_rating        11123 non-null float64
isbn                  11123 non-null object
isbn13                11123 non-null int64
language_code         11123 non-null object
  num_pages           11123 non-null int64
ratings_count         11123 non-null int64
text_reviews_count    11123 non-null int64
publication_date      11123 non-null object
publisher             11123 non-null object
dtypes: float64(1), int64(5), object(6)
memory usage: 1.0+ MB


 **We get the info that:-**
   * bookID ,isbn13 ,num_pages ,ratings_count ,text_reviews_count are of datatype int64
   * title ,authors ,isbn ,language_code ,publication_date ,publisher  are of datatype object
   * average_rating are of dataype float64
   
   ~Also the memory required for the books dataframe is about 1.0+ MB
   


**To make our data small and fast we need to decrease the memory space required by the data**
So for that we need to know what exact memory space taken by each column and the exact space of the dataframe.

## 2).Making dataframe faster and smaller

In [9]:
#To get exact memory space books dataframe is using
books.info(memory_usage = "deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11123 entries, 0 to 11122
Data columns (total 12 columns):
bookID                11123 non-null int64
title                 11123 non-null object
authors               11123 non-null object
average_rating        11123 non-null float64
isbn                  11123 non-null object
isbn13                11123 non-null int64
language_code         11123 non-null object
  num_pages           11123 non-null int64
ratings_count         11123 non-null int64
text_reviews_count    11123 non-null int64
publication_date      11123 non-null object
publisher             11123 non-null object
dtypes: float64(1), int64(5), object(6)
memory usage: 5.2 MB


~The total memory usage is 5.2MB

In [10]:
#to get the memory usage of each individual column
books.memory_usage(deep = True)

Index                      80
bookID                  88984
title                 1044646
authors                925543
average_rating          88984
isbn                   745240
isbn13                  88984
language_code          670637
  num_pages             88984
ratings_count           88984
text_reviews_count      88984
publication_date       731048
publisher              806283
dtype: int64

**Making the books dataframe fast and small**
* Converting publication_date from object to datetime datatype
* Converting language_code from object to category 

In [11]:
#changing publication_date column from object to dateTime 
books["publication_date"] = pd.to_datetime(books["publication_date"],  errors='coerce')

In [12]:
#changing language_code column from object to category
books["language_code"] = books["language_code"].astype("category")

In [13]:
books["language_code"].cat.codes.head()

0    5
1    5
2    5
3    5
4    5
dtype: int8

**why we can remove isbn column from books dataframe?**

There are 2 column isbn and isbn13 in books dataframe.isbn consist of 11 digit number assigned to books.But after 2007 13 digit number are assigned to books .Each book has a unique isbn13 and isbn 11. As both are acting as an identity for the books so we can remove one column i.e isbn column because now isbn13 is used.

In [14]:
#dropping the isbn column.
books.drop(["isbn"],axis = 1,inplace = True)

In [15]:
books.info(memory_usage = "deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11123 entries, 0 to 11122
Data columns (total 11 columns):
bookID                11123 non-null int64
title                 11123 non-null object
authors               11123 non-null object
average_rating        11123 non-null float64
isbn13                11123 non-null int64
language_code         11123 non-null category
  num_pages           11123 non-null int64
ratings_count         11123 non-null int64
text_reviews_count    11123 non-null int64
publication_date      11121 non-null datetime64[ns]
publisher             11123 non-null object
dtypes: category(1), datetime64[ns](1), float64(1), int64(5), object(3)
memory usage: 3.3 MB


Now we can see that the memory usage is 3.3 MB from 5.2 MB. 
this is beneficial cause 
* will make books dataframe faster.
* will become smaller

In [16]:
print("The dataframe contains books published from {} to {} year".format(books["publication_date"].min().year,
                                                                         books["publication_date"].max().year))


The dataframe contains books published from 1900 to 2020 year


## 3).Data Analysis and Visualization


when exploring the data there are many rows which do not have books ,so remove those rows.

In [22]:
no_book = books[books["authors"] == "NOT A BOOK"]
no_book["title"]

1061    Murder by Moonlight & Other Mysteries (New Adv...
1064    The Unfortunate Tobacconist & Other Mysteries ...
5474    The Goon Show  Volume 4: My Knees Have Fallen ...
5475               The Goon Show: Moriarty Where Are You?
5476    The Goon Show  Volume 11: He's Fallen in the W...
Name: title, dtype: object

The above list are not the books but the series so we have to delete those specific rows

In [23]:
books.drop(no_book.index,axis = 0,inplace = True)

~First analysing with respect to average rating

In [26]:
#displaying list of 5 best rated books from 1900 to 2020.
books.sort_values("average_rating",ascending = False)[["average_rating","title","authors"]].head(10)

average_rating                                              title  \
8544             5.0       Fanning the Flame: Bible  Cross  and Mission   
855              5.0  Literature Circle Guide: Bridge to Terabithia:...   
6184             5.0                          Taxation of Mineral Rents   
6775             5.0  Delwau Duon: Peintiadau Nicholas Evans = Symph...   
5648             5.0  Colossians and Philemon: A Critical and Exeget...   
5647             5.0                                Winchester Shotguns   
5023             5.0  The Complete Theory Fun Factory: Music Theory ...   
4933             5.0  Bulgakov's the Master and Margarita: The Text ...   
4788             5.0  The Diamond Color Meditation: Color Pathway to...   
4125             5.0  Zone of the Enders: The 2nd Runner Official St...   

                                               authors  
8544  Chris    Green/Chris Wright/Paul Douglas Gardner  
855                                     Tara MacCarthy  
6184                                      Ross Garnaut  
6775                   Nicholas    Evans/Rhonda  Evans  
5648                                    R. McL. Wilson  
5647                          Dennis Adler/R.L. Wilson  
5023                   Ian        Martin/Katie Elliott  
4933                                   Elena N. Mahlow  
4788                                     John  Diamond  
4125                                        Tim Bogenn

~Hence the Top average rated book is 

<img src="Images/book3.jpg">

**Now Finding the best top average_rated books for 1900s and 2000s**

In [24]:
#best top 10 average_rated of 1900s 
ts = pd.to_datetime("1/1/2000")
books19s = books.loc[books["publication_date"] < ts, : ]
books19s.sort_values("average_rating",ascending = False).head(10)

bookID                                              title  \
4933    17830  Bulgakov's the Master and Margarita: The Text ...   
624      2034  Comoediae 1: Acharenses/Equites/Nubes/Vespae/P...   
6775    25426  Delwau Duon: Peintiadau Nicholas Evans = Symph...   
6184    23276                          Taxation of Mineral Rents   
6973    26410  Wissenschaft der Logik: Die Lehre Vom Begriff ...   
6589    24814      It's a Magical World (Calvin and Hobbes  #11)   
6590    24816  Homicidal Psycho Jungle Cat (Calvin and Hobbes...   
10937   44826  The Price of the Ticket: Collected Nonfiction ...   
6591    24818                           The Days Are Just Packed   
2248     8048                                          Bill Buzz   

                                   authors  average_rating         isbn13  \
4933                       Elena N. Mahlow            5.00  9780533017423   
624    Aristophanes/F.W. Hall/W.M. Geldart            5.00  9780198145042   
6775       Nicholas    Evans/Rhonda  Evans            5.00  9780862431358   
6184                          Ross Garnaut            5.00  9780198284543   
6973         Georg Wilhelm Friedrich Hegel            4.78  9783787307678   
6589                        Bill Watterson            4.76  9780836221367   
6590                        Bill Watterson            4.72  9780836217698   
10937                        James Baldwin            4.70  9780312643065   
6591                        Bill Watterson            4.69  9780836217353   
2248        Roger Hargreaves/Gray Jolliffe            4.67  9780679801153   

      language_code    num_pages  ratings_count  text_reviews_count  \
4933            eng          202              4                   0   
624             grc          364              0                   0   
6775            wel          150              1                   0   
6184            eng          350              1                   0   
6973            ger          337              0                   0   
6589            eng          176          23875                 303   
6590            eng          176          15365                 290   
10937           eng          712            404                  30   
6591            eng          176          20308                 244   
2248            eng           32              2                   0   

      publication_date                       publisher  
4933        1975-01-01                   Vantage Press  
624         1922-02-22    Oxford University Press  USA  
6775        1987-06-22                         Y Lolfa  
6184        1983-11-17    Oxford University Press  USA  
6973        1994-01-01                       F. Meiner  
6589        1996-09-01       Andrews McMeel Publishing  
6590        1994-09-06       Andrews McMeel Publishing  
10937       1985-09-15              St. Martin's Press  
6591        1993-09-01       Andrews McMeel Publishing  
2248        1989-10-14  Random House for Young Readers

~The above list is the best 10 top average_rated books of 1900s

In [25]:
books20s = books.loc[books["publication_date"] >= ts, : ]
books20s.sort_values("average_rating",ascending = False)[["average_rating","title","authors"]].head(10)

average_rating                                              title  \
5648              5.0  Colossians and Philemon: A Critical and Exeget...   
9720              5.0      The Irish Anatomist: A Study of Flann O'Brien   
9282              5.0  Oliver Wendell Holmes in Paris: Medicine  Theo...   
9324              5.0                 Tyrannosaurus Wrecks (Stanley  #1)   
10262             5.0         Bill Gates: Computer Legend (Famous Lives)   
1243              5.0  Middlesex Borough (Images of America: New Jersey)   
855               5.0  Literature Circle Guide: Bridge to Terabithia:...   
5647              5.0                                Winchester Shotguns   
786               5.0                  Willem de Kooning: Late Paintings   
6247              5.0                        The New Big Book of America   

                                                 authors  
5648                                      R. McL. Wilson  
9720                                       Keith Donohue  
9282                                  William C. Dowling  
9324   Laura Driscoll/Alisa Klayman-Grodsky/Eric     ...  
10262                                   Sara Barton-Wood  
1243                Middlesex Borough Heritage Committee  
855                                       Tara MacCarthy  
5647                            Dennis Adler/R.L. Wilson  
786                      Julie Sylvester/David Sylvester  
6247                                Todd Davis/Marc Frey

~The above is the best 10 top average_rated books of 2000s 

ShubhangiDabral13...